rust_src/src/editfns.rs

```diff
@@ -1393,4 +1395,120 @@ pub fn delete_and_extract_region(
+    let tb = unsafe { lisp_time_struct(b, &mut blen) };
+    let t = op(ta, tb);
+    if LispObject::fixnum_overflow(t.hi) {
+        time_overflow();
+    }
+
+    let maxlen = max(alen, blen) as usize;
+    let mut v = Vec::with_capacity(maxlen);
+
+    if maxlen >= 2 {
+        v.push(LispNumber::Fixnum(t.hi));
+        v.push(LispNumber::Fixnum(t.lo.into()));
+    }
+    if maxlen >= 3 {
+        v.push(LispNumber::Fixnum(t.us.into()));
+    }
+    if maxlen > 3 {
+        v.push(LispNumber::Fixnum(t.ps.into()));
+    }
+
+    v
+}
+
+fn time_add(ta: lisp_time, tb: lisp_time) -> lisp_time {
+    let mut hi = EmacsInt::from(ta.hi + tb.hi);
+    let mut lo = ta.lo + tb.lo;
+    let mut us = ta.us + tb.us;
+    let mut ps = ta.ps + tb.ps;
+
+    if ps > 1_000_000 {
+        ps += 1;
+        us -= 1_000_000;
+    }
+    if us > 1_000_000 {
+        lo += 1;
+        us -= 1_000_000;
+    }
+    if lo > 1 << LO_TIME_BITS {
+        hi += 1;
+        lo -= 1 << LO_TIME_BITS;
+    }
+
+    lisp_time { hi, lo, us, ps }
+}
+
+fn time_subtract(ta: lisp_time, tb: lisp_time) -> lisp_time {
+    let mut hi = EmacsInt::from(ta.hi - tb.hi);
+    let mut lo = ta.lo - tb.lo;
+    let mut us = ta.us - tb.us;
+    let mut ps = ta.ps - tb.ps;
+
+    if ps < 0 {
+        us -= 1;
+        ps += 1_000_000;
+    }
+    if us < 0 {
+        lo -= 1;
+        us += 1_000_000;
+    }
+    if hi < 0 {
+        hi -= 1;
+        lo += 1 << LO_TIME_BITS;
+    }
+
+    lisp_time { hi, lo, us, ps }
+}
+
+/// Return the sum of two time values A and B, as a time value. A nil value for either argument
+/// stands for the current time. See `current-time-string' for the various forms of a time value.
+#[lisp_fn(name = "time-add", c_name = "time_add")]
+pub fn time_add_lisp(a: LispObject, b: LispObject) -> Vec<LispNumber> {
+    time_arith(a, b, time_add)
+}
+
+/// Return the difference between two time values A and B, as a time value. Use `float-time' to
+/// convert the difference into elapsed seconds.  A nil value for either argument stands for the
+/// current time.  See `current-time-string' for the various forms of a time value.
+#[lisp_fn(name = "time-subtract", c_name = "time_subtract")]
+pub fn time_subtract_lisp(a: LispObject, b: LispObject) -> Vec<LispNumber> {
+    time_arith(a, b, time_subtract)
+}
+
+macro_rules! return_if_different {
+    ($a:expr, $b:expr) => {{
+        if $a != $b {
+            return $a < $b;
+        }
+    }};
+}
+
+/// Return non-nil if time value T1 is earlier than time value T2.  A nil value for either
+/// argument stands for the current time.  See `current-time-string' for the various forms of a
+/// time value.
+#[lisp_fn]
+pub fn time_less_p(t1: LispObject, t2: LispObject) -> bool {
+    let mut t1len: c_int = 0;
+    let mut t2len: c_int = 0;
+    let a = unsafe { lisp_time_struct(t1, &mut t1len) };
+    let b = unsafe { lisp_time_struct(t2, &mut t2len) };
+
+    return_if_different!(a.hi, b.hi);
```

Comment:
Could this whole thing be simplified to `a.hi < b.hi || ... || a.ps < b.ps`?

Reply:
The original C says "a != b && a < b". Maybe to optimize the == being a faster check? 🤷‍♂️